include/openssl/opensslv.h

```diff
@@ -10,92 +10,144 @@
+# else
+#  define OPENSSL_VERSION_PRE_RELEASE_STR ""
+# endif
+# ifdef OPENSSL_VERSION_BUILD_METADATA
+#  define OPENSSL_VERSION_BUILD_METADATA_STR OPENSSL_VERSION_BUILD_METADATA
+# else
+#  define OPENSSL_VERSION_BUILD_METADATA_STR ""
+# endif
+
+/*
+ * Shared library version
  *
- * So, here's the way it works here: first of all, the library version
- * number doesn't need at all to match the overall OpenSSL version.
- * However, it's nice and more understandable if it actually does.
- * The current library version is stored in the macro SHLIB_VERSION_NUMBER,
- * which is just a piece of text in the format "M.m.e" (Major, minor, edit).
- * For the sake of Tru64, IRIX, and any other OS that behaves in similar ways,
- * we need to keep a history of version numbers, which is done in the
- * macro SHLIB_VERSION_HISTORY.  The numbers are separated by colons and
- * should only keep the versions that are binary compatible with the current.
+ * This is strictly to express ABI version, which may or may not
+ * be related to the API version expressed with the macros above.
+ * This is defined in free form.
+ */
+# define OPENSSL_SHLIB_VERSION 3
+
+/*
+ * SECTION 2: USEFUL MACROS AND FUNCTIONS
+ */
+
+/* For checking general API compatibility when preprocessing */
+# define OPENSSL_VERSION_PREREQ(maj,min)                                \
+    ((OPENSSL_VERSION_MAJOR << 16) + OPENSSL_VERSION_MINOR >= (maj << 16) + min)
+
+/*
+ * These return the values of OPENSSL_VERSION_MAJOR, OPENSSL_VERSION_MINOR,
+ * OPENSSL_VERSION_PATCH, OPENSSL_VERSION_PRE_RELEASE and
+ * OPENSSL_VERSION_BUILD_METADATA, respectively.
+ */
+unsigned int OPENSSL_version_major(void);
+unsigned int OPENSSL_version_minor(void);
+unsigned int OPENSSL_version_patch(void);
+const char *OPENSSL_version_pre_release(void);
+const char *OPENSSL_version_build_metadata(void);
+
+/*
+ * Returns the string "MAJOR.MINOR.PATCH", where MAJOR, MINOR and PATCH
+ * are replaced with the values from the corresponding OPENSSL_VERSION_
+ * macros.
+ */
+const char *OPENSSL_version_text(void);
+/*
+ * Returns the string that concatenates what OPENSSL_version_text returns
+ * together with OPENSSL_VERSION_PRE_RELEASE and OPENSSL_VERSION_BUILD_METADATA.
  */
-# define SHLIB_VERSION_HISTORY ""
-# define SHLIB_VERSION_NUMBER "1.1"
+const char *OPENSSL_version_text_full(void);
+
+/*
+ * SECTION 3: ADDITIONAL METADATA
+ */
+# define OPENSSL_RELEASE_DATE "xx XXX xxxx"
+
+/*
+ * SECTION 3: BACKWARD COMPATIBILITY
+ */
+# include <openssl/opensslconf.h>
+
+# ifndef OPENSSL_API_4
+#  ifdef OPENSSL_NO_NEWVERSION_ONLY
+/* Helper functions */
+#   define OPENSSL_MSTR_HELPER(x) #x
+#   define OPENSSL_MSTR(x) OPENSSL_MSTR_HELPER(x)
 
+/* Synthesize OPENSSL_VERSION_NUMBER with the layout 0xMNN00PPSL */
+#   define OPENSSL_VERSION_NUMBER        \
+    (long)( (OPENSSL_VERSION_MAJOR<<28)  \
+            |(OPENSSL_VERSION_MINOR<<20) \
+            |(OPENSSL_VERSION_PATCH<<4)  \
+            |0x0 )
```

Comment:
Should this last digit have a different value depending on whether `OPENSSL_VERSION_PRE_RELEASE` is defined? i..e if `OPENSSL_VERSION_PRE_RELEASE` is defined then it has the value `0x0`, otherwise it has the value `0xf`

Reply:
I dunno.  I was hoping that would not be significant, but maybe it is.  What's your opinion?

Reply:
Okie